design/cassandra/cluster-creation.md

```diff
@@ -0,0 +1,242 @@
+
+``` yaml
+apiVersion: "cassandra.rook.io/v1alpha1"
+kind: "Cluster"
+metadata:
+  name: "my-cassandra-cluster"
+spec:
+  version: "3.1.11"
+  # Optional: repository overrides the default image repo
+  repository: "custom-enterprise-repo.io/cassandra"
+  dataCenter:
+    name: "us-east-1"
+    racks:
+    - name: "us-east-1c"
+      instances: 3
+      # Optional: configMapName references a user's custom configuration for
+      # a specific Cassandra Rack 
+      configMapName: "cassandra-config"
+      # Rook Common Type: StorageSpec
+      storage:
+        volumeClaims:
+          - storageClassName:
+            metadata:
+              name: "cassandra-data"
+            spec:
+              accessModes: ["ReadWriteOnce"]
+              storageClassName: default
+              resources:
+                requests:
+                  storage: "500Gi"
+      # Rook Common Type: PlacementSpec
+      # Optional: Placement declares node/pod (anti)affinity
+      placement:
+        nodeAffinity: 
+          requiredDuringSchedulingIgnoredDuringExecution:
+            nodeSelectorTerms:
+              - matchExpressions:
+                - key: failure-domain.beta.kubernetes.io/region
+                  operator: In
+                  values:
+                  - us-east-1
+                - key: failure-domain.beta.kubernetes.io/zone
+                  operator: In
+                  values:
+                  - us-east-1c
+      # Rook Common Type: ResourceSpec
+      # Resources declares the CPU,RAM resources of a single instance of the Rack.
+      resources:
+        - name: cassandra 
+          requests:
+            cpu: "2000m"
+            memory: "4Gi"
+          limits:
+            cpu: "2000m"
+            memory: "4Gi"
+        - name: sidecar
+          requests:
+            cpu: "200m"
+            memory: "200M"
+          limits:
+            cpu: "200m"
+            memory: "200M"
+```
+
+2. **Controller** is informed of a new Cluster CRD instance. For each (dc,rack) it creates:
+    1. A StatefulSet (see Appendix A) 
+    2. A Headless Service (needed for the StatefulSet)
+    3. ClusterIP Services that serve as static IPs for Pods. Labeled accordingly if they are seeds. Each Service is named after the Pod that uses it.
+
+3. **StatefulSet Pod** starts and it starts the Cassandra and Sidecar containers. Then the following is done to customize config options for the instance:
+    1. **Cassandra**: copy default config files from `/etc/cassandra` to `/mnt/shared/default`.
+    2. **Sidecar**: Check for already created `ClusterIP` service by the Pod Name. Note if you are a Seed in the ClusterIP Service. Poll for the `default` folder, then merge default configuration with the operator's options (see Appendix B) and the user provided options. Write the result under `/mnt/shared/custom`
+    3. **Cassandra**: poll for the `custom` folder, then copy custom config files to `/etc/cassandra` and start the instance.
```

Comment:
is this an atomic operation? how does the cassandra container know that the custom folder is ready to be consumed?

Reply:
It is done atomically in the implementation.
Before the sidecar writes `custom`, it creates `custom-temp`. When `custom-temp` is ready, it renames it to `custom` with the `mv` command, which is atomic.